catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/fragment/SpriteListFragmentExplanationTextProjectWithObjectTest.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Catroid: An on-device visual programming system for Android devices
+ * Copyright (C) 2010-2018 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.uiespresso.ui.fragment;
+
+import android.support.test.runner.AndroidJUnit4;
+
+import org.catrobat.catroid.R;
+import org.catrobat.catroid.ui.ProjectActivity;
+import org.catrobat.catroid.uiespresso.content.brick.utils.BrickTestUtils;
+import org.catrobat.catroid.uiespresso.ui.fragment.rvutils.RecyclerViewActions;
+import org.catrobat.catroid.uiespresso.util.rules.BaseActivityInstrumentationRule;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import static android.support.test.espresso.Espresso.onView;
+import static android.support.test.espresso.action.ViewActions.click;
+import static android.support.test.espresso.assertion.ViewAssertions.matches;
+import static android.support.test.espresso.matcher.ViewMatchers.isDisplayed;
+import static android.support.test.espresso.matcher.ViewMatchers.withId;
+import static android.support.test.espresso.matcher.ViewMatchers.withText;
+
+import static org.catrobat.catroid.uiespresso.ui.fragment.rvutils.RecyclerViewInteractionWrapper.onRecyclerView;
+import static org.hamcrest.Matchers.allOf;
+import static org.hamcrest.Matchers.not;
+
+@RunWith(AndroidJUnit4.class)
+public class SpriteListFragmentExplanationTextProjectWithObjectTest {
+	@Rule
+	public BaseActivityInstrumentationRule<ProjectActivity> baseActivityTestRule = new
```

Comment:
Please initialize the rule with the Fragment you want to test

Reply:
e.g. CopySoundTest 
`@Rule 
public BaseActivityInstrumentationRule<SpriteActivity> baseActivityTestRule = new
			BaseActivityInstrumentationRule<>(SpriteActivity.class, SpriteActivity.EXTRA_FRAGMENT_POSITION,
			SpriteActivity.FRAGMENT_SOUNDS);`

Reply:
![screenshot_20180906-230307](https://user-images.githubusercontent.com/18437036/45185069-9e133280-b231-11e8-8904-1c9b62f23208.jpg)
I don't understand why? I'm testing this activity "see the attached screenshot" 
which is "ProjectActivity" or am I wrong?

Reply:
No, the screenshot shows the spriteListFragment inside the projectActivity. Theres are more fragments.

Just starting the activity and relying on it to show the right fragment is something i want to always prevent in the espresso test suit.

Let me explain...
The fragment stack is not cleared when a test ends, that leads to the fact that the next test starting the same activity will continue in the same fragment as the test before that left it.
Interestingly, when a test crashes, that most of the times clears the fragments again.

So for example:
We have a lot of brick tests testing the script fragment in sprite activity. Its its default fragment, so if you dont specify any it will start there.
Someone writes a test that tests the sound fragment, also inside the sprite activity. Everything works fine, unless something changes the order in which tests are run, and now the sound fragment test runs inbetween. It will cause the following scriptlist brick to fail, and only this one, (and there can be other working tests inbetween as long as they test other activities) because it ends up in sound fragment. It does this not by giving any useful error message to make things worse, but a "could not click on location 537x366" or something similar useless espresso madness.
Now add to this that per default we dont really log in which order tests are running on jenkins, especially not working tests, while here the sound fragment test is green and working, but breaking another one.

This took me days and weeks to figure out, and ive got a few more grey hairs just because of this espresso "feature".
So i want to prevent this from ever happening again, by just always explicitely stating the fragment under test, because then the issue does not happen. (Exceptions are of course fragmentless activities like stage)

Reply:
Yes, that's right, you're testing the ProjectActivity, so the code sample with the SpriteActivity was probably misleading. However, the point is, that you can also pass the Fragment you're testing explicitly as a param to the Rule, so the line should be 

```public BaseActivityInstrumentationRule<ProjectActivity> baseActivityTestRule = new BaseActivityInstrumentationRule<>(ProjectActivity.class, ProjectActivity.EXTRA_FRAGMENT_POSITION, ProjectActivity.FRAGMENT_SPRITES);```

Reply:
@LPeteR90  @84n4n4 
 **I got it. Thanks for helping me ^_***